From 42aff7a840ee9fa751dba6463fc320a5c4431489 Mon Sep 17 00:00:00 2001
From: Dave Young <dyoung@redhat.com>
Date: Thu, 22 May 2014 02:43:17 -0400
Subject: [x86] setup: Reserve setup_data ranges late after parsing memmap cmdline

Message-id: <1400726607-13717-26-git-send-email-dyoung@redhat.com>
Patchwork-id: 80126
O-Subject: [RHEL7.1 PATCH 25/35] x86: Reserve setup_data ranges late after parsing memmap cmdline
Bugzilla: 1080109
RH-Acked-by: Chao WANG <chaowang@redhat.com>
RH-Acked-by: Vivek Goyal <vgoyal@redhat.com>

Resolves: bz1080109

Brew build:
https://brewweb.devel.redhat.com/taskinfo?taskID=7480539

Test: Tested kexec/kdump on HP workstation and qemu.

Upstream commit:
mmit 77ea8c948953a90401e436e7c05973b2d5529804
Author: Dave Young <dyoung@redhat.com>
Date:   Fri Dec 20 18:02:22 2013 +0800

    x86: Reserve setup_data ranges late after parsing memmap cmdline

    Currently e820_reserve_setup_data() is called before parsing early
    params, it works in normal case. But for memmap=exactmap, the final
    memory ranges are created after parsing memmap= cmdline params, so the
    previous e820_reserve_setup_data() has no effect. For example,
    setup_data ranges will still be marked as normal system ram, thus when
    later sysfs driver ioremap them kernel will warn about mapping normal
    ram.

    This patch fix it by moving the e820_reserve_setup_data() callback after
    parsing early params so they can be set as reserved ranges and later
    ioremap will be fine with it.

    Signed-off-by: Dave Young <dyoung@redhat.com>
    Acked-by: Borislav Petkov <bp@suse.de>
    Tested-by: Toshi Kani <toshi.kani@hp.com>
    Signed-off-by: Matt Fleming <matt.fleming@intel.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/setup.c b/arch/x86/kernel/setup.c
index c22f861..961d33d 100644
--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -996,8 +996,6 @@ void __init setup_arch(char **cmdline_p)
  iomem_resource.end = (1ULL << boot_cpu_data.x86_phys_bits) - 1;
  setup_memory_map();
  parse_setup_data();
- /* update the e820_saved too */
- e820_reserve_setup_data();
 
  copy_edd();
 
@@ -1059,6 +1057,8 @@ void __init setup_arch(char **cmdline_p)
   early_dump_pci_devices();
 #endif
 
+ /* update the e820_saved too */
+ e820_reserve_setup_data();
  finish_e820_parsing();
 
  if (efi_enabled(EFI_BOOT))
-- 
1.7.1